internal/app/apid/pkg/director/director.go

```diff
@@ -0,0 +1,122 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package director provides proxy call routing facility
+package director
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/talos-systems/grpc-proxy/proxy"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Router wraps grpc-proxy StreamDirector
+type Router struct {
+	localBackends        map[string]proxy.Backend
+	remoteBackendFactory RemoteBackendFactory
+	streamedMatchers     []*regexp.Regexp
+}
+
+// RemoteBackendFactory provides backend generation by address (target)
+type RemoteBackendFactory func(target string) (proxy.Backend, error)
+
+// NewRouter builds new Router
+func NewRouter(backendFactory RemoteBackendFactory) *Router {
+	return &Router{
+		localBackends:        map[string]proxy.Backend{},
+		remoteBackendFactory: backendFactory,
+	}
+}
+
+// Register is no-op to implement factory.Registrator interface.
+//
+// Actual proxy handler is installed via grpc.UnknownServiceHandler option.
+func (r *Router) Register(srv *grpc.Server) {
+}
+
+// Director implements proxy.StreamDirector function
+func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mode, []proxy.Backend, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return r.localDirector(fullMethodName)
+	}
+
+	if _, exists := md["proxyfrom"]; exists {
+		return r.localDirector(fullMethodName)
+	}
+
+	var targets []string
+
+	if targets, ok = md["targets"]; !ok {
+		// send directly to local node, skips another layer of proxying
```

Comment:
this is something I would like to highlight.

In previous implementation, if `targets` MD field was missing, request was routed to another `apid` with target set as `md[:authority]`. I changed this behavior because `:authority` is kind of `Host:` HTTP header, it is set by the caller, so it should LB endpoint. So request goes like: `client -> LB -> apid [host1] -> LB -> apid [host2]`. In this example `host1` != `host2`. This seems to be a bit confusing path. And `apid` might not be able to talk back to the LB (why not?). Plus it requires valid TLS client cert to be present on `apid` instance (to talk back to itself), which might be tricky in failure modes (cert not renewed, etc). So it might block ability to access one node directly.

But proxying though once again when `targets` was not set allowed to set host metadata for the response, but it was set to `:authority` value, which once again might not be the node (it might be LB endpoint). So after this change if we do `osctl` without `-t`, `NODE:` comes back blank. If we do `osctl -t <targets>`, `NODE:` is set properly.

tl;dr

Now `apid` if `targets` is not set behaves like `osd` before - proxies directly to instances on the same node. If `targets` key is set, it forwards to `apid` instances.

Same goes with response wrapping, error handling, if `targets` is not set, errors are returned directly, response is passed as is, etc. (like `osd` before).

Thoughts?

Reply:
As we discussed on slack. I think this is a good change.